File: 2948-make-lexicographically-smallest-array-by-swapping-elements/2948-make-lexicographically-smallest-array-by-swapping-elements.cs
public class Solution {
    public int[] LexicographicallySmallestArray(int[] nums, int limit) {
        var n = nums.Length;

        // tuples of num and its index
        var numIdx = new List<(int num, int idx)>();
        for(var i = 0; i < n; i++)
            numIdx.Add((nums[i], i));
        
        // sort numIdx asc based on num value
        numIdx = numIdx.OrderBy(x => x.num).ToList();

        // group numbers into different sets, based on their difference
        var sets = new List<List<(int num, int idx)>>();
        sets.Add([numIdx[0]]);
        for (var i = 1; i < n; i++) {
            if(numIdx[i].num - numIdx[i - 1].num > limit)
                sets.Add([numIdx[i]]);      // separate set
            else
                sets[^1].Add(numIdx[i]);    // part of prev set
        }
        
        foreach(var sett in sets){
            // sort indexes of all elements in curr set
            var indexes = sett.Select(x => x.idx).ToList();
            indexes.Sort();

            // now put smallest no. at smallest index, then next no. and so on
            for(var i = 0; i < indexes.Count; i++)
                nums[indexes[i]] = sett[i].num;
        }
        
        return nums;
    }
}

File: 2948-make-lexicographically-smallest-array-by-swapping-elements/2948-make-lexicographically-smallest-array-by-swapping-elements.py
class Solution:
    def lexicographicallySmallestArray(self, nums: List[int], limit: int) -> List[int]:
        # tuples of num and its index
        numIdx = [] 
        for i, num in enumerate(nums):
            numIdx.append((num, i))
        
        # sort numIdx asc based on num value
        numIdx = sorted(numIdx, key=lambda k: k[0])

        # group numbers into different sets, based on their difference
        sets = [[numIdx[0]]]
        for i in range(1, len(nums)):
            if numIdx[i][0] - numIdx[i-1][0] > limit:
                sets.append([numIdx[i]])    # separate set
            else:
                sets[-1].append(numIdx[i])   # part of prev set
        
        for sett in sets:
            # sort indexes of all elements in curr set
            indexes = []
            for num, idx in sett: indexes.append(idx)
            indexes.sort()

            # now put smallest no. at smallest index, then next no. and so on
            for i, idx in enumerate(indexes):
                nums[idx] = sett[i][0]
        
        return nums

